rust-src/bulletproofs/src/inner_product_proof.rs

```diff
@@ -295,81 +297,60 @@ pub fn verify_inner_product<C: Curve>(
-        .plus_point(&H.mul_by_scalar(&b))
-        .plus_point(&Q.mul_by_scalar(&ab))
-        .minus_point(&sum);
-    P_prime.minus_point(&RHS).is_zero_point()
+    let H_exponents = vec![C::Scalar::one(); n];
+    // P_prime is also given directly, so set P_prime_bases to P_prime with exponent
+    // 1. Since it is assumed that P_prime_bases starts with G, H and Q, add those
+    // first and set the first 2n+1 exponents to 0.
+    let mut P_prime_bases = Vec::with_capacity(2 * n + 2);
+    P_prime_bases.extend(G_vec);
+    P_prime_bases.extend(H_vec);
+    P_prime_bases.push(*Q);
+    P_prime_bases.push(*P_prime);
+    let mut P_prime_exponents = Vec::with_capacity(2 * n + 2);
+    P_prime_exponents.extend(vec![C::Scalar::zero(); 2 * n + 1]);
+    P_prime_exponents.push(C::Scalar::one());
+
+    verify_inner_product_with_scalars(
+        transcript,
+        &H_exponents,
+        &P_prime_bases,
+        &P_prime_exponents,
+        proof,
+    )
 }
 
-/// This function is an optimized variant of the above.
-/// It is verified whether `P'=<a,G>+<b,H'>+<a,b>Q` for `P' =
-/// multiexp(P_prime_bases, P_prime_exponents)` and `H'_i =
-/// H_i^H_exponents_i`.
+/// This function is the actual verification function used at the core of the
+/// function `verify_inner_product`. It is verified whether
+/// `P'=<a,G>+<b,H'>+<a,b>Q` for `P' = multiexp(P_prime_bases,
+/// P_prime_exponents)` and `H'_i = H_i^H_exponents_i`.
 ///
 /// Arguments:
 /// - `transcript` - the proof transcript
-/// - `gens` - generators containing vectors `G` and `H` both of length at least
-///   `n`
 /// - `H_exponents` - slice of scalars to whose powers the `H_i` are raised
 /// - `P_prime_bases` - slice of points for computing curve point `P'`. It is
-///   assumed that the first base points are `G | H, Q`, which are implicit and
-///   omitted from `P_prime_bases`
+///   assumed that the first base points are `G | H, Q`
 /// - `P_prime_exponents` - slice of scalars to whose powers the elements
 ///   `P_prime_bases` are raised
-/// - `Q` - the elliptic curve point `Q`
 /// - `proof` - the inner product proof
 ///
 /// Preconditions:
 /// - `H_exponents` must have length `n`, which is a power of 2.
-/// - `gens` contain at least `n` pairs of generators.
+/// - `P_prime_bases` contains `G` and `H`, each consisting of `n` curve points,
+///   followed by a curve point `Q`, and may contain additional curve points.
 /// - The length of `P_prime_exponents` is equal to the length of
-///   `P_prime_bases` plus of `2n + 1` (since `G`, `H`, and `Q` are
-/// omitted from `P_prime_bases`).
+///   `P_prime_bases`
 #[allow(non_snake_case)]
 pub(crate) fn verify_inner_product_with_scalars<C: Curve>(
     transcript: &mut RandomOracle,
-    gens: &Generators<C>,
     H_exponents: &[C::Scalar],
     P_prime_bases: &[C],
     P_prime_exponents: &[C::Scalar],
-    Q: &C,
     proof: &InnerProductProof<C>,
 ) -> bool {
-    let n = gens.G_H.len();
+    let n = H_exponents.len();
```

Comment:
This seems like a significant change. Why was it previously not an issue? Do we always supply generators of the right size here?

Reply:
It was an issue before. It was undetected because all tests supplied generators of the right size (and this function was only used for set-membership, which is not used in practice, yet), but it is not a precondition. We should probably add an explicit test case for having additional generators.

Reply:
Could you add a small test for that in this PR?

Reply:
Done.